core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -274,62 +327,133 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
+      case (z, Then(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
+      case (z, Both(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
+      case (z, Traced(cause, _))  => cause.foldLeft(z)(f)
+      case (z, Meta(cause, _))    => cause.foldLeft(z)(f)
+
+      case (z, _) => z
+    }
 }
 
 object Cause extends Serializable {
 
-  final def die(defect: Throwable): Cause[Nothing]               = Die(defect)
-  final def fail[E](error: E): Cause[E]                          = Fail(error)
-  final val interrupt: Cause[Nothing]                            = Interrupt
-  final def traced[E](cause: Cause[E], trace: ZTrace): Traced[E] = Traced(cause, trace)
+  final def die(defect: Throwable): Cause[Nothing]              = Die(defect)
+  final def fail[E](error: E): Cause[E]                         = Fail(error)
+  final val interrupt: Cause[Nothing]                           = Interrupt
+  final def stack[E](cause: Cause[E]): Cause[E]                 = Meta(cause, Data(false))
+  final def stackless[E](cause: Cause[E]): Cause[E]             = Meta(cause, Data(true))
+  final def traced[E](cause: Cause[E], trace: ZTrace): Cause[E] = Traced(cause, trace)
+
+  /**
+   * Converts the specified `Cause[Option[E]]` to an `Option[Cause[E]]` by
+   * recursively stripping out any failures with the error `None`.
+   */
+  final def sequenceCauseOption[E](c: Cause[Option[E]]): Option[Cause[E]] =
+    c match {
+      case Cause.Traced(cause, trace) => sequenceCauseOption(cause).map(Cause.Traced(_, trace))
+      case Cause.Meta(cause, data)    => sequenceCauseOption(cause).map(Cause.Meta(_, data))
+      case Cause.Interrupt            => Some(Cause.Interrupt)
+      case d @ Cause.Die(_)           => Some(d)
+      case Cause.Fail(Some(e))        => Some(Cause.Fail(e))
+      case Cause.Fail(None)           => None
+      case Cause.Then(left, right) =>
+        (sequenceCauseOption(left), sequenceCauseOption(right)) match {
+          case (Some(cl), Some(cr)) => Some(Cause.Then(cl, cr))
+          case (None, Some(cr))     => Some(cr)
+          case (Some(cl), None)     => Some(cl)
+          case (None, None)         => None
+        }
+
+      case Cause.Both(left, right) =>
+        (sequenceCauseOption(left), sequenceCauseOption(right)) match {
+          case (Some(cl), Some(cr)) => Some(Cause.Both(cl, cr))
+          case (None, Some(cr))     => Some(cr)
+          case (Some(cl), None)     => Some(cl)
+          case (None, None)         => None
+        }
+    }
 
-  final case class Fail[E](value: E) extends Cause[E] {
+  private case class Fail[E](value: E) extends Cause[E] {
     override final def equals(that: Any): Boolean = that match {
       case fail: Fail[_]     => value == fail.value
       case traced: Traced[_] => this == traced.cause
+      case meta: Meta[_]     => this == meta.cause
       case _                 => false
     }
   }
 
-  final case class Die(value: Throwable) extends Cause[Nothing] {
+  object Fail {
```

Comment:
I'd make all the companion objects private, too.

Reply:
We need these if we want to expose the `apply` methods on them. Normally I wouldn't do this but there is a good amount of code that uses them, and `then` is a reserved keyword so we can't use the normal lower case smart constructor for that. I could delete the companion objects and replace with methods with capitalized names or we could go through and change everything, maybe using `parallel` and `sequential` for the smart constructors, though that is wordier.